Sources/TensorFlow/Layers/Normalization.swift

```diff
@@ -224,3 +244,166 @@ public struct LayerNorm<Scalar: TensorFlowFloatingPoint>: Layer {
+    precondition(axis != 0, "The axis cannot be batch axis.")
+    precondition(offset.rank == 1, "The offset must have rank 1.")
+    precondition(scale.rank == 1, "The scale must have rank 1.")
+    precondition(
+      offset.shape == scale.shape,
+      "The offset and the scale must have same shape.")
+    self.offset = offset
+    self.scale = scale
+    self.groups = groups
+    self.axis = axis
+    self.epsilon = epsilon
+  }
+
+  /// Creates a group normalization layer.
+  ///
+  /// - Parameters:
+  ///   - featureCount: The number of features.
+  ///   - groups: The number of groups.
+  ///   - axis: The axis that should be normalized.
+  ///   - epsilon: The small scalar added to variance.
+  public init(
+    featureCount: Int,
+    groups: Int,
+    axis: Int = -1,
+    epsilon: Scalar = 1e-3
+  ) {
+    precondition(
+      featureCount.isMultiple(of: groups),
+      "The feature count must be divisible by groups.")
+    self.init(
+      offset: Tensor(zeros: [featureCount]),
+      scale: Tensor(ones: [featureCount]),
+      groups: groups,
+      axis: axis,
+      epsilon: epsilon
+    )
+  }
+
+  /// Returns the output obtained from applying the layer to the given input.
+  ///
+  /// - Parameter input: The input to the layer.
+  /// - Returns: The output.
+  @differentiable
+  public func callAsFunction(_ input: Tensor<Scalar>) -> Tensor<Scalar> {
+    let positiveAxis = (input.rank + axis) % input.rank
+    precondition(positiveAxis != 0, "The axis cannot be batch axis.")
+    var offset = self.offset
+    var scale = self.scale
+    var broadcastShape = TensorShape([Int](repeating: 1, count: input.rank + 1))
+    broadcastShape[positiveAxis] = groups
+    broadcastShape[positiveAxis + 1] = input.shape[positiveAxis] / groups
+    offset = offset.reshaped(to: broadcastShape)
+    scale = scale.reshaped(to: broadcastShape)
+
+    var groupShape = input.shape
+    groupShape[positiveAxis] /= groups
+    groupShape.insert(groups, at: positiveAxis)
+    let grouped = input.reshaped(to: groupShape)
+    var normalizedAxes = Array(1..<grouped.rank)
+    normalizedAxes.remove(at: positiveAxis - 1)
+    let moments = grouped.moments(alongAxes: normalizedAxes)
+    let eps = Tensor(epsilon)
+    let normalized = _batchNorm(
+      grouped,
+      mean: moments.mean, variance: moments.variance,
+      offset: offset, scale: scale,
+      varianceEpsilon: eps)
+    return normalized.reshaped(to: input.shape)
+  }
+}
+
+/// A layer that applies instance normalization over a mini-batch of inputs.
+///
+/// Reference: [Instance Normalization: The Missing Ingredient for Fast Stylization](https://arxiv.org/abs/1607.08022).
+@frozen
+public struct InstanceNorm<Scalar: TensorFlowFloatingPoint>: Layer {
+  /// Internal group normalization.
+  var groupNorm: GroupNorm<Scalar>
+
+  /// The offset value, also known as beta.
+  public var offset: Tensor<Scalar> {
+    _read { yield groupNorm.offset }
+    _modify { yield &groupNorm.offset }
```

Comment:
Have we verified that this code's performance benefits from using the unofficial language `_read`/`_modify`/`yield` language features, and that a simple getter/setter pair generates worse code?

Same applies to `scale`.

Reply:
If it is `get`/`set`, the code `layer.scale[0] += 1` will be equivalent to:

```swift
var scale = layer.scale
scale[0] += 1
layer.scale = scale
```

because `Tensor` is value type.
There will be CoW and `scale` can have many elements.
With `_read`/`_modify` there'll be no CoW.

Reply:
By golly, you're right.  I thought they had fixed the optimizer for inlinable getters/setters but I just did some experiments that indicate otherwise.  Thanks for the insight!